### ISSUE TITLE
require is not defined in ES module scope

### ISSUE THREAD

**Issue by Hsinky** [caoxiemeihao/nuxt-electron#68]
when run `nuxt dev && electron .`
*error*
require is not defined in ES module scope, you can use import instead
This file is being treated as an ES module because it has a '.js' file extension and '[path]\package.json' contains "type": "module". To treat it as a CommonJS script, rename it to use the '.cjs' file extension.

*help me,please!*

**Comment by Hsinky** [comment]
It worked ! when i remove `"type": "module"` in package.json，but，I want to use es6，how to update?

**Comment by Hsinky** [comment]
> > It worked ! when i remove `"type": "module"` in package.json，but，I want to use es6，how to update?
> 
> `~/tsconfig.json`
> 
> ```
> {
> ...
> "compilerOptions": {
>   "module: "es6" 
>   // or target: "es6"
> }
> }
> ```
> 
> Maybe?

**Thanks for your help! but ,Still the same error**

**Comment by angelhdzmultimedia** [comment]
By the way, I never run `nuxt dev && electron .`. I only run `nuxt dev` and I get the electron window.

**Comment by Hsinky** [comment]
![image](https://github.com/caoxiemeihao/nuxt-electron/assets/10111222/53dcd896-6188-4885-9118-f9fedde7107f)

**Comment by angelhdzmultimedia** [comment]
Can you tell more about your environment?
Are you using JavaScript with `require`?

**Comment by Hsinky** [comment]
> Can you tell more about your environment? Are you using JavaScript with `require`?

![image](https://github.com/caoxiemeihao/nuxt-electron/assets/10111222/c527fcad-1d4e-49cb-a485-5a2ec1dd16e4)
`nuxt-electron`

**Comment by angelhdzmultimedia** [comment]
**I'm not from the team by the way 😁**

Just wanted to help. I'm starting with Electron and Nuxt, had an issue with the preload, and solved it.

All I've read on the internet mentions removing "type": "module" from package.json to get it working.

Both Nuxt and Electron are ESM ready: 
https://www.electronjs.org/docs/latest/tutorial/esm
https://nuxt.com/docs/guide/concepts/esm

So, do you want to get rid of the error, or you want to have your .js files with ESM imports instead of require?

(will be playing with tsconfig.json to see if I find something)

**Comment by Hsinky** [comment]
> **I'm not from the team by the way 😁**
> 
> Just wanted to help. I'm starting with Electron and Nuxt, had an issue with the preload, and solved it.
> 
> All I've read on the internet mentions removing "type": "module" from package.json to get it working.
> 
> Both Nuxt and Electron are ESM ready: https://www.electronjs.org/docs/latest/tutorial/esm https://nuxt.com/docs/guide/concepts/esm
> 
> So, do you want to get rid of the error, or you want to have your .js files with ESM imports instead of require?
> 
> (will be playing with tsconfig.json to see if I find something)

**Thanks for your help again!I wrote the code in nuxt-electron(https://nuxt.com/modules/electron)**

**Comment by angelhdzmultimedia** [comment]
Hmmm... 🤔
https://github.com/electron-vite/vite-plugin-electron/issues/186

**Comment by angelhdzmultimedia** [comment]
### Update
I'm getting somewhere... 👀

![image](https://github.com/caoxiemeihao/nuxt-electron/assets/10191085/f8dcebef-332e-4c1b-8542-195dc7d06ddf)

**Comment by angelhdzmultimedia** [comment]
Stay tuned to https://github.com/electron-vite/vite-plugin-electron/issues/186

**Comment by Hsinky** [comment]
> I'm getting somewhere

**My question is same to you,So,Did you solve it？**

**Comment by angelhdzmultimedia** [comment]
> > I'm getting somewhere
> 
> **My question is same to you,So,Did you solve it？**

Not yet. Will keep playing with the `renderer` options until I solve it and will let you know.

**Comment by Hsinky** [comment]
> Not yet. Will keep playing with the `renderer` options until I solve it and will let you know.

**Thanks!!""

**Comment by angelhdzmultimedia** [comment]
### Solution
Update `vite-plugin-electron` to `0.28.1`, and add `"type": "module"` in `package.json`.

**Comment by TristanDuck** [comment]
Tried all the suggestions here but still getting this:
![image](https://github.com/caoxiemeihao/nuxt-electron/assets/8880898/a047248b-406b-438c-9634-83d1cb165292)

**Comment by angelhdzmultimedia** [comment]
> vite-plugin-electron

Show package.json and nuxt.config please.

**Comment by angelhdzmultimedia** [comment]
> Tried all the suggestions here but still getting this: ![image](https://private-user-images.githubusercontent.com/8880898/306839057-a047248b-406b-438c-9634-83d1cb165292.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1NzM2ODEsIm5iZiI6MTcwODU3MzM4MSwicGF0aCI6Ii84ODgwODk4LzMwNjgzOTA1Ny1hMDQ3MjQ4Yi00MDZiLTQzOGMtOTYzNC04M2QxY2IxNjUyOTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjJUMDM0MzAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2U0ZWRlZjIzMWNhZGVkOTZiNmJiYzI1ZGExOTg0ZWVmYzNhYjlhOTI4YTU2MGVlYTIwZDk0NzM2ZDAzOGE5YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.g2Wf551IeSBKz1EMu1kNV2JPyD_KCrL3fzHDvAT4Yes)

Show package.json and nuxt.config please.

**Comment by TristanDuck** [comment]
> > Tried all the suggestions here but still getting this: ![image](https://private-user-images.githubusercontent.com/8880898/306839057-a047248b-406b-438c-9634-83d1cb165292.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1NzM2ODEsIm5iZiI6MTcwODU3MzM4MSwicGF0aCI6Ii84ODgwODk4LzMwNjgzOTA1Ny1hMDQ3MjQ4Yi00MDZiLTQzOGMtOTYzNC04M2QxY2IxNjUyOTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjJUMDM0MzAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2U0ZWRlZjIzMWNhZGVkOTZiNmJiYzI1ZGExOTg0ZWVmYzNhYjlhOTI4YTU2MGVlYTIwZDk0NzM2ZDAzOGE5YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.g2Wf551IeSBKz1EMu1kNV2JPyD_KCrL3fzHDvAT4Yes)
> 
> Show package.json and nuxt.config please.

```
// https://nuxt.com/docs/api/configuration/nuxt-config
import vuetify from 'vite-plugin-vuetify'

export default defineNuxtConfig({
  typescript: {
    shim: false,
  },
  ssr: false,
  electron: {
    build: [
      {
        // Main-Process entry file of the Electron App.
        entry: 'electron/main.ts',
      },
      {
        entry: 'electron/preload.ts',
        onstart(args) {
          // Notify the Renderer-Process to reload the page when the Preload-Scripts build is complete,
          // instead of restarting the entire Electron App.
          args.reload()
        },
      },
    ],
    renderer: {},
  },
  modules: [
    'nuxt-electron',
    (_options, nuxt) => {
      nuxt.hooks.hook('vite:extendConfig', (config) => {
        config.plugins.push(vuetify({ autoImport: true }))
      })
    },
  ],
  build: {
    transpile: ['vuetify'],
  },
  devtools: { enabled: true },
})
```
```
{
  "name": "nuxt-app",
  "version": "1.0.0",
  "private": true,
  "main": "dist-electron/main.js",
  "type": "module",
  "scripts": {
    "dev": "nuxi dev",
    "build": "nuxi build --prerender && electron-builder",
    "lint": "eslint .",
    "lint:fix": "eslint . --fix"
  },
  "dependencies": {
    "@mdi/font": "^7.4.47"
  },
  "devDependencies": {
    "@antfu/eslint-config": "^2.6.4",
    "electron": "^28.2.4",
    "electron-builder": "^24.12.0",
    "eslint": "^8.56.0",
    "nuxt": "^3.10.1",
    "nuxt-electron": "^0.7.0",
    "sass": "^1.70.0",
    "vite-plugin-electron": "0.28.1",
    "vite-plugin-electron-renderer": "^0.14.5",
    "vite-plugin-vuetify": "^2.0.1",
    "vuetify": "^3.5.3"
  }
}
```

**Comment by angelhdzmultimedia** [comment]
> > > Tried all the suggestions here but still getting this: ![image](https://private-user-images.githubusercontent.com/8880898/306839057-a047248b-406b-438c-9634-83d1cb165292.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1NzM2ODEsIm5iZiI6MTcwODU3MzM4MSwicGF0aCI6Ii84ODgwODk4LzMwNjgzOTA1Ny1hMDQ3MjQ4Yi00MDZiLTQzOGMtOTYzNC04M2QxY2IxNjUyOTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjJUMDM0MzAxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2U0ZWRlZjIzMWNhZGVkOTZiNmJiYzI1ZGExOTg0ZWVmYzNhYjlhOTI4YTU2MGVlYTIwZDk0NzM2ZDAzOGE5YSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.g2Wf551IeSBKz1EMu1kNV2JPyD_KCrL3fzHDvAT4Yes)
> > 
> > 
> > Show package.json and nuxt.config please.
> 
> ```
> // https://nuxt.com/docs/api/configuration/nuxt-config
> import vuetify from 'vite-plugin-vuetify'
> 
> export default defineNuxtConfig({
>   typescript: {
>     shim: false,
>   },
>   ssr: false,
>   electron: {
>     build: [
>       {
>         // Main-Process entry file of the Electron App.
>         entry: 'electron/main.ts',
>       },
>       {
>         entry: 'electron/preload.ts',
>         onstart(args) {
>           // Notify the Renderer-Process to reload the page when the Preload-Scripts build is complete,
>           // instead of restarting the entire Electron App.
>           args.reload()
>         },
>       },
>     ],
>     renderer: {},
>   },
>   modules: [
>     'nuxt-electron',
>     (_options, nuxt) => {
>       nuxt.hooks.hook('vite:extendConfig', (config) => {
>         config.plugins.push(vuetify({ autoImport: true }))
>       })
>     },
>   ],
>   build: {
>     transpile: ['vuetify'],
>   },
>   devtools: { enabled: true },
> })
> ```
> 
> ```
> {
>   "name": "nuxt-app",
>   "version": "1.0.0",
>   "private": true,
>   "main": "dist-electron/main.js",
>   "type": "module",
>   "scripts": {
>     "dev": "nuxi dev",
>     "build": "nuxi build --prerender && electron-builder",
>     "lint": "eslint .",
>     "lint:fix": "eslint . --fix"
>   },
>   "dependencies": {
>     "@mdi/font": "^7.4.47"
>   },
>   "devDependencies": {
>     "@antfu/eslint-config": "^2.6.4",
>     "electron": "^28.2.4",
>     "electron-builder": "^24.12.0",
>     "eslint": "^8.56.0",
>     "nuxt": "^3.10.1",
>     "nuxt-electron": "^0.7.0",
>     "sass": "^1.70.0",
>     "vite-plugin-electron": "0.28.1",
>     "vite-plugin-electron-renderer": "^0.14.5",
>     "vite-plugin-vuetify": "^2.0.1",
>     "vuetify": "^3.5.3"
>   }
> }
> ```

Please, check your `electron/main.ts` that matches this:
https://github.com/caoxiemeihao/nuxt-electron/issues/68#issuecomment-1935280744

**Comment by angelhdzmultimedia** [comment]
> Tried all the suggestions here but still getting this: ![image](https://private-user-images.githubusercontent.com/8880898/306839057-a047248b-406b-438c-9634-83d1cb165292.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1OTc4ODQsIm5iZiI6MTcwODU5NzU4NCwicGF0aCI6Ii84ODgwODk4LzMwNjgzOTA1Ny1hMDQ3MjQ4Yi00MDZiLTQzOGMtOTYzNC04M2QxY2IxNjUyOTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjJUMTAyNjI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGU5NGU2OWI4NjE4YTNmYzA2OTQ5MzFkMGI3MGIyYjUxZDU4Y2FiZGYzNDM1MzAxNzY4ODBlNWVlODVjZjljZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.nowim2bFLZP0SdP4EbAZPldLQxbmZtpeJIQ6tiiP5GA)

### SOLUTION

In `~/electron/main.ts`, replace `process.env.ROOT = path.join(process.cwd(), '..')` or `process.env.ROOT = path.join(__dirname, '..')`
with `process.env.ROOT = path.join(import.meta.url, '..')`.

As per the error shown, `__dirname` doesn't work in ES6 modules.
Replacement for modules is `import.meta.url` to get the path to the `electron/main.ts` file (to itself) and from there, 2 levels up to reach the CWD (Current Working Directory, the root of the nuxt project).

**Comment by TristanDuck** [comment]
> > Tried all the suggestions here but still getting this: ![image](https://private-user-images.githubusercontent.com/8880898/306839057-a047248b-406b-438c-9634-83d1cb165292.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg1OTc4ODQsIm5iZiI6MTcwODU5NzU4NCwicGF0aCI6Ii84ODgwODk4LzMwNjgzOTA1Ny1hMDQ3MjQ4Yi00MDZiLTQzOGMtOTYzNC04M2QxY2IxNjUyOTIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjJUMTAyNjI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGU5NGU2OWI4NjE4YTNmYzA2OTQ5MzFkMGI3MGIyYjUxZDU4Y2FiZGYzNDM1MzAxNzY4ODBlNWVlODVjZjljZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.nowim2bFLZP0SdP4EbAZPldLQxbmZtpeJIQ6tiiP5GA)
> 
> ### SOLUTION
> In `~/electron/main.ts`, replace `process.env.ROOT = path.join(process.cwd(), '..')` or `process.env.ROOT = path.join(__dirname, '..')` with `process.env.ROOT = path.join(import.meta.url, '..')`.
> 
> As per the error shown, `__dirname` doesn't work in ES6 modules. Replacement for modules is `import.meta.url` to get the path to the `electron/main.ts` file (to itself) and from there, 2 levels up to reach the CWD (Current Working Directory, the root of the nuxt project).

This worked! Thanks for the help.

**Comment by Hsinky** [comment]
> ### Solution
> Update `vite-plugin-electron` to `0.28.1`, and add `"type": "module"` in `package.json`.
> 
> **Edit:** You'll get an error because of `__dirname` in `electron/main` when defining `process.env.ROOT`. Replace with `process.cwd()`:
> 
> ```ts
> process.env.ROOT = path.join(process.cwd(), '..')
> process.env.DIST = path.join(process.env.ROOT, 'dist-electron')
> process.env.VITE_PUBLIC = process.env.VITE_DEV_SERVER_URL
>   ? path.join(process.env.ROOT, 'public')
>   : path.join(process.env.ROOT, '.output/public')
> process.env['ELECTRON_DISABLE_SECURITY_WARNINGS'] = 'true'
> const preload = path.join(process.env.DIST, 'preload.js')
> ```
**Thanks!!! There are other mistakes.**

**Comment by angelhdzmultimedia** [comment]
> > ### Solution
> > Update `vite-plugin-electron` to `0.28.1`, and add `"type": "module"` in `package.json`.
> > **Edit:** You'll get an error because of `__dirname` in `electron/main` when defining `process.env.ROOT`. Replace with `process.cwd()`:
> > ```ts
> > process.env.ROOT = path.join(process.cwd(), '..')
> > process.env.DIST = path.join(process.env.ROOT, 'dist-electron')
> > process.env.VITE_PUBLIC = process.env.VITE_DEV_SERVER_URL
> >   ? path.join(process.env.ROOT, 'public')
> >   : path.join(process.env.ROOT, '.output/public')
> > process.env['ELECTRON_DISABLE_SECURITY_WARNINGS'] = 'true'
> > const preload = path.join(process.env.DIST, 'preload.js')
> > ```
> 
> **Thanks!!! There are other mistakes.** ![image](https://private-user-images.githubusercontent.com/10111222/307713109-dda1420d-694f-4049-a051-f418532cb722.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5MzYxMjUsIm5iZiI6MTcwODkzNTgyNSwicGF0aCI6Ii8xMDExMTIyMi8zMDc3MTMxMDktZGRhMTQyMGQtNjk0Zi00MDQ5LWEwNTEtZjQxODUzMmNiNzIyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI2VDA4MjM0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTY5MjI4NmM1ZmEzMTc1NGU1NTZmMjEwMWU2MThkZDMzMjRmOGFlYTE1MWRjMmY2ZjFlY2E5ZDljNTc0ZTZiNDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GIZvVBjQezp7SGtxFeDUbvDHAU754dA7owaNjWQSDqs)

Your remove the `..` from `../preload.js`

**Comment by Hsinky** [comment]
it does't work
![image](https://github.com/caoxiemeihao/nuxt-electron/assets/10111222/814596ff-2050-4097-84c6-486f16abc72a)

**Comment by angelhdzmultimedia** [comment]
> it does't work ![image](https://private-user-images.githubusercontent.com/10111222/307715546-814596ff-2050-4097-84c6-486f16abc72a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5MzcxODksIm5iZiI6MTcwODkzNjg4OSwicGF0aCI6Ii8xMDExMTIyMi8zMDc3MTU1NDYtODE0NTk2ZmYtMjA1MC00MDk3LTg0YzYtNDg2ZjE2YWJjNzJhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI2VDA4NDEyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzZWQ2NTgwNjU0OTRhYzE1YjM1MGQzNzZiYmRiODE4ODdjNjhmYzJhOTAxOThjNDVlZTc5NDcwMjNlYTZkM2UmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.16UJqm0SOL-jc02zoORCuvdqDpxtZD3-j6rjX6oZ9C8)

`const preload = path.join(process.env.DIST, 'preload.js')`

**Comment by Hsinky** [comment]
That's not the problem.The following code runs in the development environment
```
process.env.ROOT = path.join(process.cwd())//, '..'
process.env.DIST = path.join(process.env.ROOT, 'dist-electron')
process.env.VITE_PUBLIC = process.env.VITE_DEV_SERVER_URL
  ? path.join(process.env.ROOT, 'public')
  : path.join(process.env.ROOT, '.output/public')
process.env['ELECTRON_DISABLE_SECURITY_WARNINGS'] = 'true'
const preload = path.join(process.env.DIST, 'preload.js')
```
but,It will not work after installation
![image](https://github.com/caoxiemeihao/nuxt-electron/assets/10111222/89c1e502-9a79-4e94-9d0e-225b1a98053c)

**Comment by angelhdzmultimedia** [comment]
> That's not the problem.The following code runs in the development environment
> 
> ```
> process.env.ROOT = path.join(process.cwd())//, '..'
> process.env.DIST = path.join(process.env.ROOT, 'dist-electron')
> process.env.VITE_PUBLIC = process.env.VITE_DEV_SERVER_URL
>   ? path.join(process.env.ROOT, 'public')
>   : path.join(process.env.ROOT, '.output/public')
> process.env['ELECTRON_DISABLE_SECURITY_WARNINGS'] = 'true'
> const preload = path.join(process.env.DIST, 'preload.js')
> ```
> 
> but,It will not work after installation ![image](https://private-user-images.githubusercontent.com/10111222/307732171-89c1e502-9a79-4e94-9d0e-225b1a98053c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5NDA1OTQsIm5iZiI6MTcwODk0MDI5NCwicGF0aCI6Ii8xMDExMTIyMi8zMDc3MzIxNzEtODljMWU1MDItOWE3OS00ZTk0LTlkMGUtMjI1YjFhOTgwNTNjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI2VDA5MzgxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFkM2ZmZTQ1MDhkOWIxYmFmMTdiMzVhMGUwNWViNTY0NjczYWZiOTZiMzZiNzY4NjBiMjcxNGY3ZjVlNWQzZDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RpeXVdBJeGQBr4b96BJ1HifHdNVEMwTBMTg8r8Kncyo)

That's a different issue. We are talking about ESM.

Probably you need to check if your environment is DEV or PROD and use the appropriate path like `__dirname` instead of `import.meta.url`.

**Comment by Hsinky** [comment]
**type:module in package.json,error!**(I'm not good at English.)
![image](https://github.com/caoxiemeihao/nuxt-electron/assets/10111222/8ceaf836-fd56-46dd-bfab-8790b1209b59)

**Comment by salazarr-js** [comment]
If we're trying to use [ES Modules (ESM) in Electron](https://www.electronjs.org/docs/latest/tutorial/esm) we should have **`"type": "module"`** in our `package.json` and the [preload scripts must have the .mjs extension](https://www.electronjs.org/docs/latest/tutorial/esm#esm-preload-scripts-must-have-the-mjs-extension)

The problem is that the `nuxt-electron` plugin does't do this automatically despite it using `vite-plugin-electron` internally and this was discussed/fixed on [electron-vite/vite-plugin-electron#186](https://github.com/electron-vite/vite-plugin-electron/issues/186)

### This is a little `workaround` that worked for me

#### `nuxt.config.ts` 
```ts
export default defineNuxtConfig({
  ...,
  modules: ['nuxt-electron'],
  electron: {
    renderer: { },
    build: [
      { entry: 'electron/main.ts' },
      {
        entry: 'electron/preload.ts',
        vite: {
          build: {
            rollupOptions: {
              output: {
                entryFileNames: `[name].mjs`,
                chunkFileNames: `[name].mjs`,
              },
            },
          },
        },
        onstart(args) {
          args.reload()
        },
      },
    ],
  },
})
```

This generates the proper **`dist-electron/preload.mjs`** file that the main electron script expects.

#### `electron/main.ts`
```ts
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { env } from 'node:process'
//
import { BrowserWindow, app } from 'electron'

/** Env */
const isDev = !!env.VITE_DEV_SERVER_URL
const filename = fileURLToPath(import.meta.url)
const dirname = path.dirname(filename)

env.ROOT = path.join(dirname, '../')
env.DIST = path.join(env.ROOT, 'dist-electron')
env.VITE_PUBLIC = isDev ? path.join(env.ROOT, 'public') : path.join(env.ROOT, '.output/public')
// env['ELECTRON_DISABLE_SECURITY_WARNINGS'] = 'true'

/** Create main windows and load app */
function bootstrap() {
  const preload = path.join(env.DIST!, 'preload.mjs')

  const mainWindow = new BrowserWindow({
    width: 1280,
    height: 1024,
    webPreferences: {
      preload,
      nodeIntegrationInWorker: true,
      contextIsolation: false,
      nodeIntegration: true,
      webSecurity: false, // TODO: fix `Insecure Content-Security-Policy`
    },
  })

  if (isDev) {
    mainWindow.loadURL(env.VITE_DEV_SERVER_URL!)
    mainWindow.webContents.openDevTools()
  }
  else {
    mainWindow.loadFile(path.join(env.VITE_PUBLIC!, 'index.html'))
  }
}

/** Main process */
app.whenReady().then(() => {
  bootstrap()
})
```

I've to mention that I don't know what I'm doing, nor am I a `Nuxt` or `Electron` expert, but all the @angelhdzmultimedia messages helped me to find a way, thanks 🙋‍♂️.

**Comment by angelhdzmultimedia** [comment]
> If we're trying to use [ES Modules (ESM) in Electron](https://www.electronjs.org/docs/latest/tutorial/esm) we should have **`"type": "module"`** in our `package.json` and the [preload scripts must have the .mjs extension](https://www.electronjs.org/docs/latest/tutorial/esm#esm-preload-scripts-must-have-the-mjs-extension)
> 
> The problem is that the `nuxt-electron` plugin does't do this automatically despite it using `vite-plugin-electron` internally and this was discussed/fixed on [electron-vite/vite-plugin-electron#186](https://github.com/electron-vite/vite-plugin-electron/issues/186)
> 
> ### This is a little `workaround` that worked for me
> #### `nuxt.config.ts`
> ```ts
> export default defineNuxtConfig({
>   ...,
>   modules: ['nuxt-electron'],
>   electron: {
>     renderer: { },
>     build: [
>       { entry: 'electron/main.ts' },
>       {
>         entry: 'electron/preload.ts',
>         vite: {
>           build: {
>             rollupOptions: {
>               output: {
>                 entryFileNames: `[name].mjs`,
>                 chunkFileNames: `[name].mjs`,
>               },
>             },
>           },
>         },
>         onstart(args) {
>           args.reload()
>         },
>       },
>     ],
>   },
> })
> ```
> 
> This generates the proper **`dist-electron/preload.mjs`** file that the main electron script expects.
> 
> #### `electron/main.ts`
> ```ts
> import path from 'node:path'
> import { fileURLToPath } from 'node:url'
> import { env } from 'node:process'
> //
> import { BrowserWindow, app } from 'electron'
> 
> /** Env */
> const isDev = !!env.VITE_DEV_SERVER_URL
> const filename = fileURLToPath(import.meta.url)
> const dirname = path.dirname(filename)
> 
> env.ROOT = path.join(dirname, '../')
> env.DIST = path.join(env.ROOT, 'dist-electron')
> env.VITE_PUBLIC = isDev ? path.join(env.ROOT, 'public') : path.join(env.ROOT, '.output/public')
> // env['ELECTRON_DISABLE_SECURITY_WARNINGS'] = 'true'
> 
> /** Create main windows and load app */
> function bootstrap() {
>   const preload = path.join(env.DIST!, 'preload.mjs')
> 
>   const mainWindow = new BrowserWindow({
>     width: 1280,
>     height: 1024,
>     webPreferences: {
>       preload,
>       nodeIntegrationInWorker: true,
>       contextIsolation: false,
>       nodeIntegration: true,
>       webSecurity: false, // TODO: fix `Insecure Content-Security-Policy`
>     },
>   })
> 
>   if (isDev) {
>     mainWindow.loadURL(env.VITE_DEV_SERVER_URL!)
>     mainWindow.webContents.openDevTools()
>   }
>   else {
>     mainWindow.loadFile(path.join(env.VITE_PUBLIC!, 'index.html'))
>   }
> }
> 
> /** Main process */
> app.whenReady().then(() => {
>   bootstrap()
> })
> ```
> 
> I've to mention that I don't know what I'm doing, nor am I a `Nuxt` or `Electron` expert, but all the @angelhdzmultimedia messages helped me to find a way, thanks 🙋‍♂️.

Hey! Thanks for the feedback. Will try your solution.

**Comment by caoxiemeihao** [comment]
**package.json**
```diff
{
- "type": "module",
}
